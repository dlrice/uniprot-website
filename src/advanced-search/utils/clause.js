import { v1 } from 'uuid';
import { serializableDeepAreEqual, removeProperty } from '../../utils/utils';

export const createEmptyClause = () => ({
  id: v1(),
  logicOperator: 'AND',
  field: {
    label: 'Any',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: 'single',
    dataType: 'string',
    id: 'id_Any',
  },
  queryInput: {},
});

// export const isClauseTouched = (clause) => {
//   const empty = createEmptyClause();
//   const emptyWithoutID = removeProperty(empty, 'id');
//   const clauseWithoutID = removeProperty(clause, 'id');
//   return !serializableDeepEquals(emptyWithoutID, clauseWithoutID);
// };

export const clausesAreEqual = (clause1, clause2) => serializableDeepAreEqual(removeProperty(clause1, 'id'), removeProperty(clause2, 'id'));
