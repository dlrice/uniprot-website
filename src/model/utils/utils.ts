/* eslint-disable @typescript-eslint/no-explicit-any */
const hasContent = (obj: any) => {
  return Object.values(obj).some(val => {
    if (Array.isArray(val)) {
      const valArray = val as any[];
      return valArray.length > 0;
    }
    if (typeof val === 'object' && val) {
      return Object.values(val).length > 0;
    }
    return false;
  });
};

export default hasContent;
