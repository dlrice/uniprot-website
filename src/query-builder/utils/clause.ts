import { v1 } from 'uuid';

import {
  Operator,
  Clause,
  ItemType,
  DataType,
  FieldType,
} from '../types/searchTypes';

export const createEmptyClause = (): Clause => ({
  id: v1(),
  logicOperator: Operator.AND,
  searchTerm: {
    label: 'All',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: ItemType.single,
    dataType: DataType.string,
    fieldType: FieldType.general,
    id: 'id_all',
  },
  queryBits: {},
});

export const createPreSelectedClauses = (): Clause[] => [
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      label: 'Gene Name [GN]',
      term: 'gene',
      example: 'ydj1',
      itemType: ItemType.single,
      dataType: DataType.string,
      fieldType: FieldType.general,
      id: 'gene_field',
    },
    queryBits: {},
  },
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      id: 'taxonomy_name',
      label: 'Taxonomy [OC]',
      itemType: ItemType.single,
      term: 'taxonomy_name',
      dataType: DataType.string,
      fieldType: FieldType.general,
      example: 'human',
      autoComplete: '/uniprot/api/suggester?dict=taxonomy&query=?',
      autoCompleteQueryTerm: 'taxonomy_id',
    },
    queryBits: {},
  },
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      id: 'keyword_field',
      label: 'Keyword [KW]',
      term: 'keyword',
      example: 'chromosomal',
      itemType: ItemType.single,
      dataType: DataType.string,
      fieldType: FieldType.general,
      autoComplete: '/uniprot/api/suggester?dict=keyword&query=?',
      autoCompleteQueryTerm: 'keyword',
    },
    queryBits: {},
  },
  createEmptyClause(),
];
