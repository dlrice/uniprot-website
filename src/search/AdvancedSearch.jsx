import React from 'react';
import ClauseList from './ClauseList';

const AdvancedSearch = (props) => {
  const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = props;
  return (
    <div className="advanced-search">
      <div>
        <label htmlFor="namespace-select">
          Searching in
          <select id="namespace-select">
            <option>{namespace}</option>
          </select>
        </label>
      </div>
      <ClauseList {...props} />
      <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
        Add Field
      </button>
      <div>
        <button
          type="button"
          id="submit-query"
          className="button"
          onClick={handleAdvancedSubmitClick}
        >
          Search
        </button>
      </div>
    </div>
  );
};

export default AdvancedSearch;
