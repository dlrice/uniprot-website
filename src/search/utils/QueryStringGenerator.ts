import { Clause } from '../types/searchTypes';

type IPrefixMap = {
  feature: string;
  comment: string;
  [key: string]: string;
};

const getItemTypePrefix = (itemType: string) => {
  const itemTypeToPrefixMap: IPrefixMap = {
    feature: 'ft_',
    comment: 'cc_',
  };
  return itemTypeToPrefixMap[itemType] || '';
};

const getItemTypeEvidencePrefix = (itemType: string) => {
  const itemTypeToEvidencePrefixMap: IPrefixMap = {
    feature: 'ftev_',
    comment: 'ccev_',
  };
  return itemTypeToEvidencePrefixMap[itemType] || '';
};

const getItemTypeRangePrefix = (itemType: string) =>
  itemType === 'feature' ? 'ftlen_' : '';

const createTermString = (
  term: string | undefined,
  itemType: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return 'ec:';
    }
    throw new Error('ID value not provided in query');
  }
  if (['organism', 'taxonomy', 'host'].includes(term)) {
    if (id) {
      return `${term}_id:`;
    }
    return `${term}_name:`;
  }
  const itemTypePrefix = getItemTypePrefix(itemType);
  return `${itemTypePrefix}${term}${term ? ':' : ''}`;
};

const createValueString = (
  term: string | undefined,
  valuePrefix: string | undefined,
  stringValue: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return id;
    }
    throw new Error('ID value not provided in query');
  }
  if (id) {
    if (['cofactor_chebi', 'catalytic_activity'].includes(term)) {
      return `"${id}"`;
    }
    if (
      ['organism', 'taxonomy', 'host', 'keyword', 'scl_term'].includes(term)
    ) {
      return id;
    }
  }

  // We are testing for term=xref and valuePrefix=any because the
  // search API expects the valuePrefix to be ommited in this case.
  // eg xref:foo rather than xref_any:foo
  const valuePrefixChecked =
    valuePrefix && !(term === 'xref' && valuePrefix === 'any')
      ? `${valuePrefix}-`
      : '';
  return `${valuePrefixChecked}${stringValue}`;
};

const createSimpleSubquery = (clause: Clause) => {
  if (clause.searchTerm.itemType === 'group') {
    throw Error('Cannot create a query with a group term.');
  }
  const { itemType, term, valuePrefix } = clause.searchTerm;
  const { stringValue, id } = clause.queryInput;
  if (!stringValue) {
    throw new Error('Value not provided in query');
  }
  if (term === 'All') {
    return stringValue;
  }
  const termString = createTermString(term, itemType, id);
  const valueString = createValueString(term, valuePrefix, stringValue, id);
  return `(${termString}${valueString})`;
};

const createRangeSubquery = (clause: Clause) => {
  const { term, itemType } = clause.searchTerm;
  const { rangeFrom, rangeTo } = clause.queryInput;
  const rangeFromChecked = rangeFrom || '';
  const rangeToChecked = rangeTo || '';
  const itemTypeRangePrefix = getItemTypeRangePrefix(itemType);
  return `(${itemTypeRangePrefix}${term}:[${rangeFromChecked} TO ${rangeToChecked}])`;
};

const getEvidenceSubquery = (clause: Clause) => {
  const { evidenceValue } = clause.queryInput;
  const { term, itemType } = clause.searchTerm;
  if (!evidenceValue) {
    throw new Error('Evidence value not provided');
  }
  const itemTypeEvidencePrefix = getItemTypeEvidencePrefix(itemType);
  return `(${itemTypeEvidencePrefix}${term}:${evidenceValue})`;
};

const createQueryString = (clauses: Clause[] = []): string =>
  clauses.reduce((queryAccumulator: string, clause: Clause) => {
    const query = [];
    if (
      (clause.queryInput.id && clause.queryInput.id !== '') ||
      (clause.queryInput.stringValue && clause.queryInput.stringValue !== '')
    ) {
      query.push(createSimpleSubquery(clause));
    }
    if (clause.queryInput.rangeFrom || clause.queryInput.rangeTo) {
      query.push(createRangeSubquery(clause));
    }
    if (
      clause.queryInput.evidenceValue &&
      clause.queryInput.evidenceValue !== ''
    ) {
      query.push(getEvidenceSubquery(clause));
    }
    let queryJoined = query.join(' AND ');
    if (query.length > 1) {
      queryJoined = `(${queryJoined})`;
    }

    let logicOperator = '';
    if (queryAccumulator.length > 0 && query.length > 0) {
      logicOperator = ` ${clause.logicOperator} `;
    } else if (
      queryAccumulator.length === 0 &&
      clause.logicOperator === 'NOT'
    ) {
      logicOperator = `${clause.logicOperator} `;
    }

    return `${queryAccumulator}${logicOperator}${queryJoined}`;
  }, '');

export default createQueryString;
