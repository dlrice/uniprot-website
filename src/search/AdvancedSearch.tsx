import React, { Component } from 'react';
import ClauseList from './ClauseList';
import { unpackQueryUrl } from '../utils/apiUrls';
import { Namespace } from './types/searchTypes';
import './styles/AdvancedSearch.scss';

type AdvancedSearchProps = {
  handleAdvancedSubmitClick: () => void;
  namespace: Namespace;
  dispatchAddClause: () => void;
};
class AdvancedSearch extends Component<AdvancedSearchProps> {
  componentDidMount() {
    const {
      searchTerms,
      dispatchUpdateClauses,
      dispatchfetchEvidencesIfNeeded,
      dispatchFetchSearchTermsIfNeeded,
      queryString,
    } = this.props;
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
    dispatchFetchSearchTermsIfNeeded();

    if (searchTerms.length && queryString) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  componentDidUpdate(prevProps) {
    const { queryString: prevQueryString, searchTerms: prevSearchTerms } = prevProps;
    const { queryString, searchTerms, dispatchUpdateClauses } = this.props;
    if (
      searchTerms.length
      && queryString
      && (queryString !== prevQueryString || searchTerms !== prevSearchTerms)
    ) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  render() {
    const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = this.props;
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        <ClauseList {...this.props} />
        <button type="button" id="add-field" onClick={dispatchAddClause}>
          Add Field
        </button>
        <div>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={handleAdvancedSubmitClick}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

export default AdvancedSearch;
