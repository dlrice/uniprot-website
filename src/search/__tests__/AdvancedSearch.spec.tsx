import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createEmptyClause } from '../utils/clause';
import AdvancedSearch from '../AdvancedSearch';

configure({ adapter: new Adapter() });

let wrapper;
let props;

describe('AdvancedSearch shallow components', () => {
  beforeEach(() => {
    props = {
      dispatchAddClause: jest.fn(),
      handleAdvancedSubmitClick: jest.fn(),
      dispatchFetchSearchTermsIfNeeded: jest.fn(),
      dispatchfetchEvidencesIfNeeded: jest.fn(),
      dispatchCopyQueryClausesToSearch: jest.fn(),
      history: {
        push: jest.fn(),
      },
      clauses: [...Array(4)].map(() => createEmptyClause()),
      namespace: 'UniProtKB',
      searchTerms: {
        data: [],
      },
      evidences: {
        go: {
          data: [],
          isFetching: false,
        },
        annotation: {
          data: [],
          isFetching: false,
        },
      },
    };
    wrapper = shallow(<AdvancedSearch {...props} />);
  });

  test('should render', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should call to get search terms', () => {
    expect(props.dispatchFetchSearchTermsIfNeeded).toHaveBeenCalled();
  });

  test('should call to get evidences', () => {
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('go');
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('annotation');
  });

  test('should add field rows', () => {
    wrapper.find('#add-field').simulate('click');
    expect(props.dispatchAddClause).toHaveBeenCalled();
  });

  test('should submit a query', () => {
    wrapper.find('#submit-query').simulate('click');
    expect(props.handleAdvancedSubmitClick).toHaveBeenCalled();
  });
});
