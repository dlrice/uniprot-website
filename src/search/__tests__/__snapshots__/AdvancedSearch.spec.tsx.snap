// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch shallow components should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdvancedSearch
    clauses={
      Array [
        Object {
          "id": 0,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 0,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 0,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 0,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
      ]
    }
    dispatchAddClause={[MockFunction]}
    dispatchCopyQueryClausesToSearch={[MockFunction]}
    dispatchFetchSearchTermsIfNeeded={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    dispatchfetchEvidencesIfNeeded={
      [MockFunction] {
        "calls": Array [
          Array [
            "go",
          ],
          Array [
            "annotation",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    evidences={
      Object {
        "annotation": Object {
          "data": Array [],
          "isFetching": false,
        },
        "go": Object {
          "data": Array [],
          "isFetching": false,
        },
      }
    }
    handleAdvancedSubmitClick={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    namespace="UniProtKB"
    searchTerms={
      Object {
        "data": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <label
            htmlFor="namespace-select"
          >
            Searching in
            <select
              id="namespace-select"
            >
              <option>
                UniProtKB
              </option>
            </select>
          </label>
        </div>,
        <ClauseList
          clauses={
            Array [
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
            ]
          }
          dispatchAddClause={[MockFunction]}
          dispatchCopyQueryClausesToSearch={[MockFunction]}
          dispatchFetchSearchTermsIfNeeded={
            [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            }
          }
          dispatchfetchEvidencesIfNeeded={
            [MockFunction] {
              "calls": Array [
                Array [
                  "go",
                ],
                Array [
                  "annotation",
                ],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            }
          }
          evidences={
            Object {
              "annotation": Object {
                "data": Array [],
                "isFetching": false,
              },
              "go": Object {
                "data": Array [],
                "isFetching": false,
              },
            }
          }
          handleAdvancedSubmitClick={[MockFunction]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          namespace="UniProtKB"
          searchTerms={
            Object {
              "data": Array [],
            }
          }
        />,
        <button
          id="add-field"
          onClick={[MockFunction]}
          type="button"
        >
          Add Field
        </button>,
        <div>
          <button
            className="button"
            id="submit-query"
            onClick={[MockFunction]}
            type="button"
          >
            Search
          </button>
        </div>,
      ],
      "className": "advanced-search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <label
            htmlFor="namespace-select"
          >
            Searching in
            <select
              id="namespace-select"
            >
              <option>
                UniProtKB
              </option>
            </select>
          </label>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Searching in",
              <select
                id="namespace-select"
              >
                <option>
                  UniProtKB
                </option>
              </select>,
            ],
            "htmlFor": "namespace-select",
          },
          "ref": null,
          "rendered": Array [
            "Searching in",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <option>
                  UniProtKB
                </option>,
                "id": "namespace-select",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "UniProtKB",
                },
                "ref": null,
                "rendered": "UniProtKB",
                "type": "option",
              },
              "type": "select",
            },
          ],
          "type": "label",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clauses": Array [
            Object {
              "id": 0,
              "logicOperator": "AND",
              "queryInput": Object {},
              "searchTerm": Object {
                "dataType": "string",
                "example": "a4_human, P05067, cdc7 human",
                "id": "id_all",
                "itemType": "single",
                "label": "All",
                "term": "All",
              },
            },
            Object {
              "id": 0,
              "logicOperator": "AND",
              "queryInput": Object {},
              "searchTerm": Object {
                "dataType": "string",
                "example": "a4_human, P05067, cdc7 human",
                "id": "id_all",
                "itemType": "single",
                "label": "All",
                "term": "All",
              },
            },
            Object {
              "id": 0,
              "logicOperator": "AND",
              "queryInput": Object {},
              "searchTerm": Object {
                "dataType": "string",
                "example": "a4_human, P05067, cdc7 human",
                "id": "id_all",
                "itemType": "single",
                "label": "All",
                "term": "All",
              },
            },
            Object {
              "id": 0,
              "logicOperator": "AND",
              "queryInput": Object {},
              "searchTerm": Object {
                "dataType": "string",
                "example": "a4_human, P05067, cdc7 human",
                "id": "id_all",
                "itemType": "single",
                "label": "All",
                "term": "All",
              },
            },
          ],
          "dispatchAddClause": [MockFunction],
          "dispatchCopyQueryClausesToSearch": [MockFunction],
          "dispatchFetchSearchTermsIfNeeded": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "dispatchfetchEvidencesIfNeeded": [MockFunction] {
            "calls": Array [
              Array [
                "go",
              ],
              Array [
                "annotation",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "evidences": Object {
            "annotation": Object {
              "data": Array [],
              "isFetching": false,
            },
            "go": Object {
              "data": Array [],
              "isFetching": false,
            },
          },
          "handleAdvancedSubmitClick": [MockFunction],
          "history": Object {
            "push": [MockFunction],
          },
          "namespace": "UniProtKB",
          "searchTerms": Object {
            "data": Array [],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add Field",
          "id": "add-field",
          "onClick": [MockFunction],
          "type": "button",
        },
        "ref": null,
        "rendered": "Add Field",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="button"
            id="submit-query"
            onClick={[MockFunction]}
            type="button"
          >
            Search
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search",
            "className": "button",
            "id": "submit-query",
            "onClick": [MockFunction],
            "type": "button",
          },
          "ref": null,
          "rendered": "Search",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <label
              htmlFor="namespace-select"
            >
              Searching in
              <select
                id="namespace-select"
              >
                <option>
                  UniProtKB
                </option>
              </select>
            </label>
          </div>,
          <ClauseList
            clauses={
              Array [
                Object {
                  "id": 0,
                  "logicOperator": "AND",
                  "queryInput": Object {},
                  "searchTerm": Object {
                    "dataType": "string",
                    "example": "a4_human, P05067, cdc7 human",
                    "id": "id_all",
                    "itemType": "single",
                    "label": "All",
                    "term": "All",
                  },
                },
                Object {
                  "id": 0,
                  "logicOperator": "AND",
                  "queryInput": Object {},
                  "searchTerm": Object {
                    "dataType": "string",
                    "example": "a4_human, P05067, cdc7 human",
                    "id": "id_all",
                    "itemType": "single",
                    "label": "All",
                    "term": "All",
                  },
                },
                Object {
                  "id": 0,
                  "logicOperator": "AND",
                  "queryInput": Object {},
                  "searchTerm": Object {
                    "dataType": "string",
                    "example": "a4_human, P05067, cdc7 human",
                    "id": "id_all",
                    "itemType": "single",
                    "label": "All",
                    "term": "All",
                  },
                },
                Object {
                  "id": 0,
                  "logicOperator": "AND",
                  "queryInput": Object {},
                  "searchTerm": Object {
                    "dataType": "string",
                    "example": "a4_human, P05067, cdc7 human",
                    "id": "id_all",
                    "itemType": "single",
                    "label": "All",
                    "term": "All",
                  },
                },
              ]
            }
            dispatchAddClause={[MockFunction]}
            dispatchCopyQueryClausesToSearch={[MockFunction]}
            dispatchFetchSearchTermsIfNeeded={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            dispatchfetchEvidencesIfNeeded={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "go",
                  ],
                  Array [
                    "annotation",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            evidences={
              Object {
                "annotation": Object {
                  "data": Array [],
                  "isFetching": false,
                },
                "go": Object {
                  "data": Array [],
                  "isFetching": false,
                },
              }
            }
            handleAdvancedSubmitClick={[MockFunction]}
            history={
              Object {
                "push": [MockFunction],
              }
            }
            namespace="UniProtKB"
            searchTerms={
              Object {
                "data": Array [],
              }
            }
          />,
          <button
            id="add-field"
            onClick={[MockFunction]}
            type="button"
          >
            Add Field
          </button>,
          <div>
            <button
              className="button"
              id="submit-query"
              onClick={[MockFunction]}
              type="button"
            >
              Search
            </button>
          </div>,
        ],
        "className": "advanced-search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <label
              htmlFor="namespace-select"
            >
              Searching in
              <select
                id="namespace-select"
              >
                <option>
                  UniProtKB
                </option>
              </select>
            </label>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Searching in",
                <select
                  id="namespace-select"
                >
                  <option>
                    UniProtKB
                  </option>
                </select>,
              ],
              "htmlFor": "namespace-select",
            },
            "ref": null,
            "rendered": Array [
              "Searching in",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <option>
                    UniProtKB
                  </option>,
                  "id": "namespace-select",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "UniProtKB",
                  },
                  "ref": null,
                  "rendered": "UniProtKB",
                  "type": "option",
                },
                "type": "select",
              },
            ],
            "type": "label",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clauses": Array [
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
              Object {
                "id": 0,
                "logicOperator": "AND",
                "queryInput": Object {},
                "searchTerm": Object {
                  "dataType": "string",
                  "example": "a4_human, P05067, cdc7 human",
                  "id": "id_all",
                  "itemType": "single",
                  "label": "All",
                  "term": "All",
                },
              },
            ],
            "dispatchAddClause": [MockFunction],
            "dispatchCopyQueryClausesToSearch": [MockFunction],
            "dispatchFetchSearchTermsIfNeeded": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            },
            "dispatchfetchEvidencesIfNeeded": [MockFunction] {
              "calls": Array [
                Array [
                  "go",
                ],
                Array [
                  "annotation",
                ],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            },
            "evidences": Object {
              "annotation": Object {
                "data": Array [],
                "isFetching": false,
              },
              "go": Object {
                "data": Array [],
                "isFetching": false,
              },
            },
            "handleAdvancedSubmitClick": [MockFunction],
            "history": Object {
              "push": [MockFunction],
            },
            "namespace": "UniProtKB",
            "searchTerms": Object {
              "data": Array [],
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Field",
            "id": "add-field",
            "onClick": [MockFunction],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add Field",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="button"
              id="submit-query"
              onClick={[MockFunction]}
              type="button"
            >
              Search
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
              "className": "button",
              "id": "submit-query",
              "onClick": [MockFunction],
              "type": "button",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
